File: ai-harness/skills/skill-builder/tools/screenshot.js
#!/usr/bin/env node
// screenshot.js — playwright self-review harness for visual artifacts.
//
// Used by skill-builder during Phase 2 (calibrate via samples) so the
// agent can iterate on a rendered artifact via screenshots before
// showing the user.
//
// Usage:
//   node tools/screenshot.js <html-file> [<html-file> ...]
//
// Each HTML file is opened in headless chromium at three viewports
// (16:9 desktop, wide desktop, portrait tablet). Screenshots land in
// /tmp/skill-screenshots/<viewport>/<basename>/.
//
// Reveal-detection: if the page exposes window.Reveal, every slide is
// screenshotted (with all fragments revealed). Otherwise a single
// full-page screenshot is taken.
//
// Requires playwright. If you don't have it installed:
//   npm install -g playwright && npx playwright install chromium

const path = require('path');
const fs = require('fs');

// Find playwright in: the script's tree, the cwd, or the first arg's directory.
// Iterating samples often happens in /tmp/<sample>/ where the user has run
// `npm install playwright` locally.
function findPlaywright() {
  const tries = [
    () => require('playwright'),
    () => require(path.join(process.cwd(), 'node_modules', 'playwright')),
  ];
  if (process.argv[2]) {
    tries.push(() => require(path.join(path.dirname(path.resolve(process.argv[2])), 'node_modules', 'playwright')));
  }
  for (const t of tries) {
    try { return t(); } catch {}
  }
  return null;
}

const pw = findPlaywright();
if (!pw) {
  console.error('playwright not installed. Install it in one of:');
  console.error('  - globally: npm install -g playwright && npx playwright install chromium');
  console.error('  - in your sample dir: cd <dir> && npm install playwright && npx playwright install chromium');
  process.exit(2);
}
const { chromium } = pw;

const VIEWPORTS = [
  { name: '16x9',     width: 1280, height: 800  },
  { name: 'wide',     width: 1920, height: 1080 },
  { name: 'portrait', width: 900,  height: 1200 },
];

const SHOTS_DIR = '/tmp/skill-screenshots';

async function shootDeck(page, dir) {
  await page.waitForFunction(() => window.Reveal && window.Reveal.isReady && window.Reveal.isReady());
  await page.waitForTimeout(400);

  const total = await page.evaluate(() => window.Reveal.getTotalSlides());

  for (let i = 0; i < total; i++) {
    await page.evaluate((idx) => window.Reveal.slide(idx), i);
    await page.evaluate(() => {
      // Reveal all fragments so the screenshot shows the slide complete
      document.querySelectorAll('.present .fragment').forEach(f => {
        f.classList.add('visible', 'current-fragment');
      });
    });
    await page.waitForTimeout(700); // chart/math render time
    const num = String(i + 1).padStart(2, '0');
    await page.screenshot({ path: path.join(dir, `slide-${num}.png`), fullPage: false });
  }
  return total;
}

async function shootPage(page, dir) {
  await page.waitForTimeout(700);
  await page.screenshot({ path: path.join(dir, 'page.png'), fullPage: true });
  return 1;
}

async function shoot(file, viewport) {
  const browser = await chromium.launch();
  const ctx = await browser.newContext({
    viewport: { width: viewport.width, height: viewport.height },
    deviceScaleFactor: 1,
  });
  const page = await ctx.newPage();

  const fileUrl = 'file://' + path.resolve(file);
  await page.goto(fileUrl, { waitUntil: 'networkidle' });

  const isDeck = await page.evaluate(() => typeof window.Reveal !== 'undefined');

  // Use parent folder name when the file is generically named (e.g. index.html).
  const stem = path.basename(file, path.extname(file));
  const base = stem === 'index' ? path.basename(path.dirname(path.resolve(file))) : stem;
  const dir = path.join(SHOTS_DIR, viewport.name, base);
  fs.mkdirSync(dir, { recursive: true });

  const n = isDeck ? await shootDeck(page, dir) : await shootPage(page, dir);

  await browser.close();
  return { n, isDeck, dir };
}

(async () => {
  const files = process.argv.slice(2);
  if (files.length === 0) {
    console.error('Usage: node screenshot.js <html-file> [<html-file> ...]');
    process.exit(2);
  }

  // Clean shots dir
  if (fs.existsSync(SHOTS_DIR)) fs.rmSync(SHOTS_DIR, { recursive: true });

  for (const file of files) {
    if (!fs.existsSync(file)) {
      console.error(`skip: ${file} not found`);
      continue;
    }
    for (const vp of VIEWPORTS) {
      const { n, isDeck, dir } = await shoot(file, vp);
      const kind = isDeck ? `${n} slides` : 'full page';
      console.log(`${path.basename(file)} ${vp.name}: ${kind} → ${dir}/`);
    }
  }
})();

File: ai-harness/skills/skill-builder/tools/screenshot.py
#!/usr/bin/env -S uv run -q --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["playwright>=1.52"]
# ///
"""Capture visual self-review screenshots for generated HTML artifacts.

Usage:
  uv run --script tools/screenshot.py <html-file> [<html-file> ...]

Each HTML file is opened in headless Chromium at three viewports. Screenshots
land in /tmp/skill-screenshots/<timestamp>-<pid>/<viewport>/<basename>/.

If the page exposes window.Reveal, every slide is screenshotted with fragments
shown. Otherwise the tool captures one full-page screenshot.
"""

from __future__ import annotations

import subprocess
import sys
from dataclasses import dataclass
from datetime import datetime, timezone
from os import getpid
from pathlib import Path
from typing import Sequence

from playwright.sync_api import (
    Browser,
    BrowserType,
    Error as PlaywrightError,
    Page,
    Playwright,
    sync_playwright,
)


SHOTS_DIR = Path("/tmp/skill-screenshots")


@dataclass(frozen=True)
class Viewport:
    name: str
    width: int
    height: int


@dataclass(frozen=True)
class ShotResult:
    count: int
    is_deck: bool
    directory: Path


VIEWPORTS = (
    Viewport("16x9", 1280, 800),
    Viewport("wide", 1920, 1080),
    Viewport("portrait", 900, 1200),
)


def browser_missing(error: PlaywrightError) -> bool:
    message = str(error)
    return "Executable doesn't exist" in message or "playwright install" in message


def install_chromium() -> None:
    print(
        "Chromium is missing; running `playwright install chromium`.", file=sys.stderr
    )
    result = subprocess.run(
        [sys.executable, "-m", "playwright", "install", "chromium"],
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError("failed to install Playwright Chromium")


def launch_chromium(chromium: BrowserType) -> Browser:
    try:
        return chromium.launch()
    except PlaywrightError as exc:
        if not browser_missing(exc):
            raise
        install_chromium()
        return chromium.launch()


def output_base_name(file: Path) -> str:
    stem = file.stem
    return file.parent.name if stem == "index" else stem


def make_run_root() -> Path:
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    return SHOTS_DIR / f"{timestamp}-{getpid()}"


def shoot_deck(page: Page, directory: Path) -> int:
    page.wait_for_function(
        "() => window.Reveal && window.Reveal.isReady && window.Reveal.isReady()"
    )
    page.wait_for_timeout(400)

    total = int(page.evaluate("() => window.Reveal.getTotalSlides()"))
    for idx in range(total):
        page.evaluate("(slideIndex) => window.Reveal.slide(slideIndex)", idx)
        page.evaluate(
            """() => {
              document.querySelectorAll(".present .fragment").forEach((fragment) => {
                fragment.classList.add("visible", "current-fragment");
              });
            }"""
        )
        page.wait_for_timeout(700)
        page.screenshot(
            path=str(directory / f"slide-{idx + 1:02d}.png"),
            full_page=False,
        )

    return total


def shoot_page(page: Page, directory: Path) -> int:
    page.wait_for_timeout(700)
    page.screenshot(path=str(directory / "page.png"), full_page=True)
    return 1


def shoot(
    playwright: Playwright,
    file: Path,
    viewport: Viewport,
    run_root: Path,
) -> ShotResult:
    browser = launch_chromium(playwright.chromium)
    try:
        context = browser.new_context(
            viewport={"width": viewport.width, "height": viewport.height},
            device_scale_factor=1,
        )
        page = context.new_page()
        page.goto(file.resolve().as_uri(), wait_until="networkidle")

        is_deck = bool(page.evaluate("() => typeof window.Reveal !== 'undefined'"))
        directory = run_root / viewport.name / output_base_name(file.resolve())
        directory.mkdir(parents=True, exist_ok=True)

        count = shoot_deck(page, directory) if is_deck else shoot_page(page, directory)
        context.close()
        return ShotResult(count=count, is_deck=is_deck, directory=directory)
    finally:
        browser.close()


def main(argv: Sequence[str]) -> int:
    files = [Path(arg) for arg in argv]
    if not files:
        print(
            "Usage: uv run --script screenshot.py <html-file> [<html-file> ...]",
            file=sys.stderr,
        )
        return 2

    run_root = make_run_root()

    with sync_playwright() as playwright:
        for file in files:
            if not file.exists() or not file.is_file():
                print(f"skip: {file} not found", file=sys.stderr)
                continue
            for viewport in VIEWPORTS:
                result = shoot(playwright, file, viewport, run_root)
                kind = f"{result.count} slides" if result.is_deck else "full page"
                print(f"{file.name} {viewport.name}: {kind} -> {result.directory}/")

    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
